File: frontend/src/types/post.ts
export interface PostResponse {
  id: string;
  numericId: number;
  title: string;
  author: string;
  content: string;
  views: number;
  comments: number;
  pinned: boolean;
  pinnedAt: string | null;
  experience: "experienced" | "inexperienced" | null;
  category: "question" | "info" | "chat" | null;
  createdAt: string;
  updatedAt: string;
}

export interface PostListResponse {
  success: boolean;
  posts: PostResponse[];
  count: number;
  total: number;
  totalPages: number;
  currentPage: number;
  limit: number;
}

export interface PostCreateRequest {
  author: string;
  title: string;
  content: string;
  password: string;
  experience?: "experienced" | "inexperienced";
  category?: "question" | "info" | "chat";
}

export interface PostUpdateRequest {
  title?: string;
  content?: string;
  password: string;
}

export interface PostDeleteRequest {
  password?: string;
  isAdmin?: boolean;
}

export interface LikeToggleRequest {
  visitorId: string;
}

export interface LikeResponse {
  success: boolean;
  liked: boolean;
  likes: number;
}

export interface LikeStatusResponse {
  success: boolean;
  likes: number;
  liked: boolean;
}

export interface BulkLikesResponse {
  success: boolean;
  likes: Record<string, number>;
}

export type PostSearchType = "title" | "author" | "content";
export type PostSortField = "newest" | "oldest" | "likes" | "views" | "comments";
export type PostCategoryFilter = "question" | "info" | "chat";

File: backend/app/schemas/post.py
from datetime import datetime
from enum import Enum
from typing import Optional, Literal

from pydantic import Field

from app.schemas.common import CamelModel


class PostSortField(str, Enum):
    NEWEST = "newest"
    OLDEST = "oldest"
    LIKES = "likes"
    VIEWS = "views"
    COMMENTS = "comments"


class PostSearchType(str, Enum):
    TITLE = "title"
    AUTHOR = "author"
    CONTENT = "content"


class PostCategoryFilter(str, Enum):
    QUESTION = "question"
    INFO = "info"
    CHAT = "chat"


class PostCreate(CamelModel):
    author: str = Field(min_length=1, max_length=50)
    title: str = Field(min_length=1, max_length=200)
    content: str = Field(min_length=1, max_length=10000)
    password: str = Field(min_length=4, max_length=50)
    experience: Optional[Literal["experienced", "inexperienced"]] = None
    category: Optional[Literal["question", "info", "chat"]] = None


class PostUpdate(CamelModel):
    title: Optional[str] = Field(default=None, min_length=1, max_length=200)
    content: Optional[str] = Field(default=None, min_length=1, max_length=10000)
    password: str = Field(min_length=4, max_length=50)


class PostResponse(CamelModel):
    id: str
    numeric_id: int
    title: str
    author: str
    content: str
    views: int
    comments: int
    pinned: bool
    pinned_at: datetime | None
    experience: str | None
    category: str | None
    created_at: datetime
    updated_at: datetime


class PostListResponse(CamelModel):
    success: bool = True
    posts: list[PostResponse]
    count: int
    total: int
    total_pages: int
    current_page: int
    limit: int


class PostCreateResponse(CamelModel):
    success: bool = True
    message: str
    post: PostResponse


class PostUpdateResponse(CamelModel):
    success: bool = True
    message: str
    post: PostResponse


class PostDeleteRequest(CamelModel):
    password: str | None = None


class PasswordVerifyRequest(CamelModel):
    password: str = Field(min_length=1)


class PasswordVerifyResponse(CamelModel):
    success: bool = True
    verified: bool


class LikeToggleRequest(CamelModel):
    visitor_id: str = Field(alias="visitorId", min_length=8)


class LikeResponse(CamelModel):
    success: bool = True
    liked: bool
    likes: int


class LikeStatusResponse(CamelModel):
    success: bool = True
    likes: int
    liked: bool


class BulkLikesRequest(CamelModel):
    post_ids: list[str] = Field(alias="postIds", max_length=100)


class BulkLikesResponse(CamelModel):
    success: bool = True
    likes: dict[str, int]


class ViewIncrementResponse(CamelModel):
    success: bool = True
    views: int


class PinToggleResponse(CamelModel):
    success: bool = True
    message: str
    post: PostResponse

